File: src/cli/highlighter.py
"""
SQL 语法高亮器
"""

import re
from typing import List, Tuple
from prompt_toolkit.formatted_text import StyleAndTextTuples
from prompt_toolkit.lexers import Lexer


# SQL 关键字分类
SQL_KEYWORDS = {
    # 子句关键字 - 蓝色
    "blue": [
        "SELECT", "FROM", "WHERE", "GROUP", "BY", "HAVING", "ORDER",
        "LIMIT", "OFFSET", "UNION", "INTERSECT", "EXCEPT",
        "INSERT", "INTO", "VALUES", "UPDATE", "DELETE", "REPLACE",
        "CREATE", "ALTER", "DROP", "TRUNCATE",
        "TABLE", "INDEX", "VIEW", "DATABASE",
        "JOIN", "INNER", "LEFT", "RIGHT", "OUTER", "CROSS", "NATURAL",
        "ON", "USING",
    ],
    # 条件关键字 - 紫色
    "purple": [
        "AND", "OR", "NOT", "IN", "EXISTS", "BETWEEN", "LIKE", "ILIKE",
        "IS", "NULL", "TRUE", "FALSE",
        "REGEXP", "RLIKE", "SIMILAR", "TO",
    ],
    # 函数和聚合 - 青色
    "cyan": [
        "COUNT", "SUM", "AVG", "MAX", "MIN", "GROUP_CONCAT",
        "CAST", "CONVERT", "COALESCE", "NULLIF",
        "DISTINCT", "ALL", "ANY", "SOME",
        "ASC", "DESC",
    ],
    # 其他关键字 - 默认色
    "default": [
        "AS", "CASE", "WHEN", "THEN", "ELSE", "END",
        "DEFAULT", "PRIMARY", "KEY", "FOREIGN", "REFERENCES",
        "CONSTRAINT", "UNIQUE", "CHECK",
        "SET", "SHOW", "DESCRIBE", "EXPLAIN", "USE",
        "COMMIT", "ROLLBACK", "START", "TRANSACTION",
        "RETURNING", "CONFLICT", "DO", "NOTHING",
    ],
}

# 构建关键字正则
def _build_keyword_pattern() -> re.Pattern:
    """构建关键字匹配正则"""
    all_keywords = []
    for color, keywords in SQL_KEYWORDS.items():
        all_keywords.extend(keywords)

    # 按长度降序排序，优先匹配长的关键字
    all_keywords.sort(key=len, reverse=True)

    # 构建正则：\b(KEYWORD)\b
    pattern = r'\b(' + '|'.join(all_keywords) + r')\b'
    return re.compile(pattern, re.IGNORECASE)


KEYWORD_PATTERN = _build_keyword_pattern()

# 字符串模式（单引号、双引号）
STRING_PATTERN = re.compile(r"('[^']*'|\"[^\"]*\")")

# 数字模式
NUMBER_PATTERN = re.compile(r'\b\d+(\.\d+)?\b')

# 注释模式（单行 -- 和多行 /* */）
COMMENT_PATTERN = re.compile(r'(--[^\n]*|/\*.*?\*/)', re.DOTALL)

# 标识符（表名、列名）- 反引号或方括号包裹
IDENTIFIER_PATTERN = re.compile(r'(`[^`]+`|\[[^\]]+\])')


class SimpleSQLLexer(Lexer):
    """简化的 SQL 语法高亮器 - 用于性能敏感场景"""

    def __init__(self):
        # 构建关键字字典（用于快速查找）
        self._keyword_styles = {}
        for color, keywords in SQL_KEYWORDS.items():
            style = f"ansiblue" if color == "blue" else \
                    "ansimagenta" if color == "purple" else \
                    "ansicyan" if color == "cyan" else "ansiwhite"
            for kw in keywords:
                self._keyword_styles[kw.upper()] = style

    def lex_document(self, document):
        """返回一个函数，该函数给定行号返回该行的样式化文本"""
        def get_line(lineno):
            text = document.lines[lineno] if lineno < len(document.lines) else ""
            return self.lex(text)
        return get_line

    def lex(self, text: str) -> StyleAndTextTuples:
        """简单的词法分析"""
        result: List[Tuple[str, str]] = []

        # 使用正则分割文本
        tokens = re.split(r"(\s+|[(),=;<>!+\-*/])", text)

        for token in tokens:
            if not token:
                continue

            # 检查是否是字符串
            if (token.startswith("'") and token.endswith("'")) or \
               (token.startswith('"') and token.endswith('"')):
                result.append(('ansiyellow', token))

            # 检查是否是数字
            elif re.match(r'^\d+(\.\d+)?$', token):
                result.append(('ansigreen', token))

            # 检查是否是注释
            elif token.startswith('--'):
                result.append(('ansilightgray italic', token))

            # 检查是否是大写关键字
            elif token.upper() in self._keyword_styles:
                result.append((self._keyword_styles[token.upper()], token))

            # 其他（普通文本）
            else:
                result.append(('', token))

        return result

File: src/ts/utils/highlighter.ts
/**
 * SQL 语法高亮工具
 * 为 SQL 语句提供简单的关键字着色
 */

import chalk from 'chalk';

// SQL 关键字列表
const SQL_KEYWORDS = new Set([
  'SELECT', 'FROM', 'WHERE', 'AND', 'OR', 'NOT', 'IN', 'IS', 'NULL',
  'LIKE', 'BETWEEN', 'EXISTS', 'CASE', 'WHEN', 'THEN', 'ELSE', 'END',
  'JOIN', 'INNER', 'LEFT', 'RIGHT', 'OUTER', 'ON', 'AS',
  'GROUP', 'BY', 'HAVING', 'ORDER', 'ASC', 'DESC', 'LIMIT', 'OFFSET',
  'UNION', 'ALL', 'INTERSECT', 'EXCEPT',
  'INSERT', 'INTO', 'VALUES', 'UPDATE', 'SET', 'DELETE',
  'CREATE', 'DROP', 'ALTER', 'TABLE', 'INDEX', 'VIEW', 'TRIGGER',
  'DATABASE', 'SCHEMA', 'PRIMARY', 'KEY', 'FOREIGN', 'REFERENCES',
  'UNIQUE', 'CHECK', 'DEFAULT', 'CONSTRAINT', 'ADD', 'MODIFY',
  'CHANGE', 'RENAME', 'COLUMN', 'IF', 'NULLS', 'FIRST', 'AFTER',
  'USE', 'SHOW', 'DESCRIBE', 'EXPLAIN', 'ANALYZE',
  'TRUE', 'FALSE', 'DISTINCT', 'COUNT', 'SUM', 'AVG', 'MIN', 'MAX',
  'CAST', 'CONVERT', 'COALESCE', 'NULLIF', 'CASE',
  'WITH', 'RECURSIVE', 'TEMP', 'TEMPORARY',
  'TRANSACTION', 'BEGIN', 'COMMIT', 'ROLLBACK',
  'INDEXED', 'CROSS', 'NATURAL', 'USING',
]);

// SQL 函数列表
const SQL_FUNCTIONS = new Set([
  'COUNT', 'SUM', 'AVG', 'MIN', 'MAX', 'ABS', 'ROUND', 'CEIL', 'FLOOR',
  'LENGTH', 'TRIM', 'UPPER', 'LOWER', 'SUBSTR', 'REPLACE',
  'DATE', 'DATETIME', 'TIME', 'YEAR', 'MONTH', 'DAY', 'HOUR', 'MINUTE', 'SECOND',
  'NOW', 'CURRENT_DATE', 'CURRENT_TIME', 'CURRENT_TIMESTAMP',
  'COALESCE', 'NULLIF', 'IFNULL', 'CAST', 'CONVERT',
  'CONCAT', 'GROUP_CONCAT', 'JSON_EXTRACT', 'JSON_OBJECT', 'JSON_ARRAY',
]);

export interface HighlightOptions {
  useColors?: boolean;
  keywordColor?: keyof typeof chalk;
  stringColor?: keyof typeof chalk;
  numberColor?: keyof typeof chalk;
  commentColor?: keyof typeof chalk;
  functionColor?: keyof typeof chalk;
}

const defaultOptions: HighlightOptions = {
  useColors: true,
  keywordColor: 'cyan',
  stringColor: 'green',
  numberColor: 'yellow',
  commentColor: 'gray',
  functionColor: 'blue',
};

/**
 * 高亮 SQL 语句
 */
export function highlightSql(sql: string, options: HighlightOptions = {}): string {
  const opts = { ...defaultOptions, ...options };

  if (!opts.useColors) {
    return sql;
  }

  let result = sql;

  // 1. 先处理多行注释 /* ... */
  const multiLineComments: string[] = [];
  result = result.replace(/\/\*[\s\S]*?\*\//g, (match) => {
    const id = `__MULTILINE_COMMENT_${multiLineComments.length}__`;
    multiLineComments.push(chalk.gray(match));
    return id;
  });

  // 2. 处理单行注释 -- ...
  const singleLineComments: string[] = [];
  result = result.replace(/--.*$/gm, (match) => {
    const id = `__SINGLELINE_COMMENT_${singleLineComments.length}__`;
    singleLineComments.push(chalk.gray(match));
    return id;
  });

  // 3. 处理字符串（单引号和双引号）
  const strings: string[] = [];
  result = result.replace(/('[^']*'|"[^"]*")/g, (match) => {
    const id = `__STRING_${strings.length}__`;
    strings.push(chalk.green(match));
    return id;
  });

  // 4. 处理数字
  const numbers: string[] = [];
  result = result.replace(/\b(\d+\.?\d*)\b/g, (match) => {
    const id = `__NUMBER_${numbers.length}__`;
    numbers.push(chalk.yellow(match));
    return id;
  });

  // 5. 处理 SQL 函数（后跟括号的标识符）
  const functions: string[] = [];
  result = result.replace(/\b([A-Za-z_][A-Za-z0-9_]*)\s*\(/g, (match, p1) => {
    const funcName = p1.toUpperCase();
    if (SQL_FUNCTIONS.has(funcName)) {
      const id = `__FUNCTION_${functions.length}__`;
      functions.push(chalk.blue(p1) + match.slice(p1.length));
      return id + '(';
    }
    return match;
  });

  // 6. 处理 SQL 关键字
  result = result.replace(/\b([A-Za-z_][A-Za-z0-9_]*)\b/g, (match) => {
    const upper = match.toUpperCase();
    if (SQL_KEYWORDS.has(upper)) {
      return chalk.cyan(match);
    }
    return match;
  });

  // 7. 恢复函数占位符
  functions.forEach((colored, i) => {
    result = result.replace(`__FUNCTION_${i}__`, colored);
  });

  // 8. 恢复数字占位符
  numbers.forEach((colored, i) => {
    result = result.replace(`__NUMBER_${i}__`, colored);
  });

  // 9. 恢复字符串占位符
  strings.forEach((colored, i) => {
    result = result.replace(`__STRING_${i}__`, colored);
  });

  // 10. 恢复注释占位符
  singleLineComments.forEach((colored, i) => {
    result = result.replace(`__SINGLELINE_COMMENT_${i}__`, colored);
  });
  multiLineComments.forEach((colored, i) => {
    result = result.replace(`__MULTILINE_COMMENT_${i}__`, colored);
  });

  return result;
}

/**
 * 高亮 SQL 语句（简化版，用于输出回显）
 */
export function highlightSqlSimple(sql: string): string {
  return highlightSql(sql, {
    useColors: true,
    keywordColor: 'cyan',
    stringColor: 'green',
    numberColor: 'yellow',
    commentColor: 'gray',
    functionColor: 'blue',
  });
}

/**
 * 移除 ANSI 颜色代码
 */
export function stripAnsi(text: string): string {
  return text.replace(/\x1b\[[0-9;]*m/g, '');
}

/**
 * 获取 SQL 语句的类型
 */
export function getSqlType(sql: string): string {
  const trimmed = sql.trim().toUpperCase();

  if (trimmed.startsWith('SELECT')) return 'SELECT';
  if (trimmed.startsWith('INSERT')) return 'INSERT';
  if (trimmed.startsWith('UPDATE')) return 'UPDATE';
  if (trimmed.startsWith('DELETE')) return 'DELETE';
  if (trimmed.startsWith('CREATE')) return 'CREATE';
  if (trimmed.startsWith('DROP')) return 'DROP';
  if (trimmed.startsWith('ALTER')) return 'ALTER';
  if (trimmed.startsWith('WITH')) return 'WITH';
  if (trimmed.startsWith('EXPLAIN')) return 'EXPLAIN';
  if (trimmed.startsWith('SHOW')) return 'SHOW';

  return 'UNKNOWN';
}
